roachtest: add tpcdsvec test

yuzefovich opened the pull request:
**roachtest: add tpcdsvec test**

This commit adds a new `tpcdsvec` test which creates a 3 node cluster,
restores TPCDS dataset for the scale factor of 1, and then runs most of
TPCDS queries on two separate connections, one with `vectorize=off` and
another with `vectorize=on`. The returned results from both connections
are compared using `cmpconn` package. The test eventually fails if the
results on one of the queries will differ between two connections.

This commit also slightly adjusts queries 71 and 79 to make them return
deterministic output by appending two and one, respectively, columns to
ORDER BY clauses.

Fixes: #45407.

Release note: None

**cmd: move cmpconn package outside of smithcmp**

This commit refactors `cmpconn` package to make it easier to use for
non-mutating cases and moves it outside of `smithcmp` package.

Release note: None

yuzefovich committed:
roachtest: add tpcdsvec test

This commit adds a new `tpcdsvec` test which creates a 3 node cluster,
restores TPCDS dataset for the scale factor of 1, and then runs most of
TPCDS queries on two separate connections, one with `vectorize=off` and
another with `vectorize=on`. The returned results from both connections
are compared using `cmpconn` package. The test eventually fails if the
results on one of the queries will differ between two connections.

This commit also slightly adjusts queries 71 and 79 to make them return
deterministic output by appending two and one, respectively, columns to
ORDER BY clauses.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47049)
<!-- Reviewable:end -->

yuzefovich committed:
cmd: move cmpconn package outside of smithcmp

This commit refactors `cmpconn` package to make it easier to use for
non-mutating cases and moves it outside of `smithcmp` package.

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47049)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @rohany)

___
*[pkg/cmd/roachtest/tpcdsvec.go, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47049#-M470MN0lWoy0oKycnr--r1-187:-M4GGUou1-ykO4DCJ9uc:b-c5i00f) ([raw file](https://github.com/cockroachdb/cockroach/blob/93e5064bcb6e805f7fcfec203603989dfadfcb54/pkg/cmd/roachtest/tpcdsvec.go#L187)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

how are you comparing vectorize = on and off times with this?
</blockquote></details>

In this test we do not do such comparison currently. The initial goal is to test correctness, and for now I don't think we're interested in performance comparison. The thing is that `CompareConns` will execute the query on both connections, and we're measuring the time for both configs of a single query.
___
*[pkg/cmd/roachtest/tpcdsvec.go, line 197 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47049#-M470MN0lWoy0oKycnr--r1-197:-M4GGp6r5YrRHum4UcH4:be8f29a) ([raw file](https://github.com/cockroachdb/cockroach/blob/93e5064bcb6e805f7fcfec203603989dfadfcb54/pkg/cmd/roachtest/tpcdsvec.go#L197)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

dont you want to know for each vectorize on and off, whether it was better with and without stats? 
</blockquote></details>

I think of this item as a nice-to-have. As I said above, we currently do not distinguish the runtimes between two options of `vectorize` setting. I think that comparison of `no stats vec OFF + no stats vec ON` vs `with stats vec OFF + with stats vec ON` will still highlight whether a worse plan is chosen when stats are present. I opened #46701 about such cases for TPCH benchmark, so I thought it'd be nice to do this automatically on TPCDS as well. I think later we will add `tpcdsvec/perf` test that will expand the comparison configurations.
___
*[pkg/cmd/smithcmp/cmpconn/conn.go, line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47049#-M470MN0lWoy0oKycnr0-r1-11:-M4GZ1oX7tF160pBusvj:b18ru34) ([raw file](https://github.com/cockroachdb/cockroach/blob/93e5064bcb6e805f7fcfec203603989dfadfcb54/pkg/cmd/smithcmp/cmpconn/conn.go#L11)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

maybe it is worth pulling this package out of smithcmp now that it is used by more than just sqlsmith
</blockquote></details>

I also was thinking this but got lazy. Your question persuaded me that it's worth it. Done.


<!-- Sent from Reviewable.io -->

rohany reviewed (approved):
understood, thanks for the refactor!

yuzefovich commented:
TFTR!

bors r+
